### PR TITLE
Don't include partial types in nav-to results that exist only to wrap another nested type.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to #50128

Before:

image

After

image

The general idea here is that users often make partial types just so they can split a nested type out into its own file.  In this case, including the containing partial type from the broken-out file in navto is really not beneficial to the user as they're not really searching for that when they put in the name.  Instead, they really want to find the primary container and go to that.  

This is an alternative/complimentary PR to https://github.com/dotnet/roslyn/pull/50139.

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs [hunk 1]
```diff
@@ -1169,6 +1167,73 @@ public void ToError(__arglist)
     VerifyNavigateToResultItem(item, "ToError", "[|ToError|](__arglist)", PatternMatchKind.Exact, NavigateToItemKind.Method, Glyph.MethodPublic);
 });
         }
+
+        [Fact]
+        public async Task DoNotIncludeTrivialPartialContainer()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs:1172]: 💡 Can you add a test where all parts are considered "trivial containers"?

File: src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs [hunk 2]
```diff
@@ -1169,6 +1167,73 @@ public void ToError(__arglist)
...
+            VerifyNavigateToResultItems(
+                new()
+                {
+                    new NavigateToItem("Outer", NavigateToItemKind.Class, "csharp", null, null, s_emptyExactPatternMatch, null),
+                },
+                await _aggregator.GetItemsAsync("Outer"));
+        }
+
+        [Fact]
+        public async Task DoIncludeNonTrivialPartialContainer()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs:1205]: 💡 Can you include a test where a part contains both a member and a nested type?

Comment (sharwell) [inline src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs:1205]: 💡 Can you include a test where a part contains multiple nested types, but no other members?

Comment (sharwell) [inline src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs:1205]: ❔ Can you describe the behavior of two-level (or more) nested types? Tests would be good for non-obvious cases.

**Comment by CyrusNajmabadi** [comment]
Sure, I'll beef up the testing to help clarify all the expected behaviors.

File: src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs:1269]: Effectively, a type is considered 'trivial' (and not included in the index) if it is partial, has members, and *all* those members are nested types.  This applies to a type regardless of if the type itself is an outer type or a nested type.

The general intuition here is that such types don't really arise in practice *except* to be a way to have nested types broken out from teh 'main' outer type.  And, as suh, we don't really consider those a valuable 'navigate to' target for a user.  in other words, given files:

Foo.cs
```cs
partial class Foo
{
     // main Foo bits
}
```

and Foo.Bar.cs:
```c#
partial class Foo
{
    class Bar
    {
        // Bar bits.
    }
}
```

The user doesn't think of the 'Foo' in Foo.Bar.cs as being significant (in terms of navigation).  It exists merely to ensure 'Bar' is properly contained.  So only navigable symbols in Foo.Bar.cs is just `class Bar` (and whatever is inside of that).

File: src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs [hunk 3]
```diff
@@ -17,7 +17,7 @@ namespace Microsoft.CodeAnalysis.FindSymbols
     internal sealed partial class SyntaxTreeIndex : IObjectWritable
     {
         private const string PersistenceName = "<SyntaxTreeIndex>";
-        private static readonly Checksum SerializationFormatChecksum = Checksum.Create("20");
+        private static readonly Checksum SerializationFormatChecksum = Checksum.Create("22");
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs:20]:
```suggestion
        private static readonly Checksum SerializationFormatChecksum = Checksum.Create("21");
```

**Comment by genlu** [comment]
From your comment in another PR, re: ```partial class Outer { class Inner { } }``` as the only declaration of `Outer`, is there a real use case for it (besides source generator)? Is it possible to have an analyzer to fade unnecessary partial that works with source generator?

**Review by genlu — APPROVED** [review]
:shipit:

File: src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs [hunk 4]
```diff
@@ -1169,6 +1167,225 @@ public void ToError(__arglist)
...
+    </Project>
+</Workspace>
+", composition: EditorTestCompositions.EditorFeatures);
+
+            _provider = new NavigateToItemProvider(workspace, AsynchronousOperationListenerProvider.NullListener);
+            _aggregator = new NavigateToTestAggregator(_provider);
+
+            VerifyNavigateToResultItems(
+                new() { },
+                await _aggregator.GetItemsAsync("Outer"));
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs:1265]: 📝 We should probably bring this one to design review

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs:1265]: Sure v will bring to next meeting

**Pull request merged**